doc/design/doi-generation-2.md

```diff
@@ -0,0 +1,212 @@
+Every public published version of a Dandiset will recieve a **Version DOI**.
+
+For example:
+Dandiset DOI: `https://doi.org/10.48324/dandi.000027/`
+Version DOI: `https://doi.org/10.48324/dandi.000027/0.210831.2033`
+
+Prior to publication, the Dandiset DOI will refer to the latest draft version.
+Following publication, the Dandiset DOI will refer to the latest published version.
+
+At creation the `Dandiset DOI` will be a DataCite `Draft DOI`, but will be "promoted" to a DataCite `Findable DOI` as soon as possible.
+`Version DOI` will always be a `Findable DOI`.
+
+## The current approach
+
+- [initial design doc](./doi-generation-1.md)
+- overall:
+   - leave DOI absent upon dandiset creation
+   - upon publication
+      - inject fake DOI (but do not save) and validate
+      - after validation, create a new `Version DOI` (function `create_doi`)
+      - publish dandiset
+
+### Issues with the Existing Approach
+
+- [Stop injecting "fake" DOIs into draft dandisets](https://github.com/dandi/dandi-archive/issues/1709)
+- [Unpublished Dandisets display a DOI under `Cite As`](https://github.com/dandi/dandi-archive/issues/1932)
+
+## Proposed Solution
+
+Initially proposed/discussed in
+
+- [Create and maintain a "Findable" DOI for the Dandiset as a whole](https://github.com/dandi/dandi-archive/issues/1319)
+
+and boils down to the adoption of approach of Zenodo of having a DOI which always points to the latest version of the record.
+[Zenodo uses the language](https://support.zenodo.org/help/en-gb/1-upload-deposit/97-what-is-doi-versioning) `Concept DOI` to mean a top-level DOI that references all versions, which we will refer to as `Dandiset DOI`.
+
+DataCite allows for three types of DOIs ([DataCite](https://support.datacite.org/docs/what-does-the-state-of-the-doi-mean)):
+
+- `Draft`. We do not use those.
+  *Can be deleted, and they require only the identifier itself in order to be created or saved. They can be updated to either Registered or Findable DOIs. Registered and Findable DOIs may not be returned to the Draft state, which means that changing the state of a Draft is final.*
+- `Registered`. Like `Findable` but not indexed for search, so we do not use them.
+- `Findable`. Is the type we use for published dandisets.
+  Requires to be valid (pass validation to fit the datacite schema) to be created.
+
+We propose to:
+
+- Upon creation of a **public** dandiset, mint a `Dandiset DOI` (a DataCite `Draft DOI`) `10.48324/dandi.{dandiset.id}` with
+  - *minimal metadata* entered during creation request (title, description, license)
+  - DLP URL `https://dandiarchive.org/dandiset/{dandiset.id}`
+  - For embargoed dandiset, **do not** specify any metadata besides the DLP URL.
+  - If minting a DOI fails, we need to raise exception to inform developers about the issue but proceed with the creation of the dandiset.
+- Upon changes to a non-embargoed, draft dandiset metadata record:
+  - If `Draft DOI`, attempt to "promote" it to `Findable`.
```

Comment:
@asmacdo @djarecka how easy/reliable to check if DOI is `Draft` or `Findable`?  

On one hand, ideally we should avoid needing to check the state on datacite since it would require extra interaction which could go bad.  On the other, this way we would not assume knowing what is the state of the DOI, and whether it was successfully produced/promoted before.  It brings possibility for a race condition, but I think the danger and possible side-effects are minimal (unless I see some big problem).

Reply:
Easy, a GET to the DOI, response["data"]["attributes"]["state"]

Reply:
note: if we do not use validated models (e.g. pydantic), it would require try/except or conversion to chain of `.get` to ensure robust operation and appropriate logic

Reply:
The validation is jsonschema based https://github.com/dandi/dandi-archive/pull/2012#issuecomment-2828205136

The only thing I think we need to worry about is if it exists or not. Whether draft or findable, I think we can do the same payload with "event=publish", pretty sure it would promote if draft or keep as findable as long as the data is valid. 

see also
https://github.com/dandi/dandi-schema/blob/master/dandischema/datacite/__init__.py#L265-L269

Reply:
(@yarikoptic this is my previously misplaced comment)
@djarecka and I discussed this and we aren't convinced it is really necessary to promote unpublished Dandisets to Findable, especially since it leads to that awkward scenario of a Findable DOI with incorrect metadata

"Adding a comment somewhere" -- Since we cant use the data on the dandiset, (it would be invalid) we could possibly retrieve the DOI, and modify that data, but this feels unclean to me.

IMO the cleaner path would be to simply leave Draft Dandisets as Draft DOIs until their first publish. At that point mint the Dandiset DOI. And only update the Dandiset DOI after further successful publishes.

Reply:
agree, let's keep Draft until publish, and after that keep updating only upon subsequent Publishes